### ISSUE TITLE
Problem connecting to GKE private clusters

### ISSUE THREAD

**Issue by bygui86** [lensapp/lens#1731]
**Describe the bug**
Till version 3.6.9 everything works perfectly but since the release of 4.0, **Lens literally does not work anymore on MacOS**.

**To Reproduce**
1. Open Lens
2. Click on any cluster
3. Click on `Nodes` or `Workloads/Deployment` or `Network/Services` or `Storage/Persistent Volumes` or anything else different than Pods...
4. See Lens loading forever without showing any info

Even loading pods logs has become impossible:
1. Open Lens
2. Click on any cluster
3. Click on `Workloads/Pods`
4. Click on any pod
5. Try to show logs
6. See Lens loading forever without showing any log

Impossible to load metrics (and yes, I have whole prometheus stack deployed and worked till version 3.6.9!!):
1. Open Lens
2. Click on any cluster
3. See Lens loading forever without showing any metrics

The left menu of CRDs does not work as before
1. Open Lens
2. Click on any cluster
3. Click on `Custom Resources`
4. All CRDs submenus are opened at once, I'm not able to close them selectively
5. Click on any CRD
5. See Lens loading forever without showing anything

**Expected behavior**
I expect that everything works same as in version 3.6.9:
- I'm able to see Nodes and any other k8s resource in the cluster
- I'm able to show pods logs
- I'm able to see metrics (cluster, pods, etc)

**Screenshots**
Loading kubernetes cluster metrics or pod metrics
![Screenshot 2020-12-10 at 09 05 58](https://user-images.githubusercontent.com/22518997/101739918-80943980-3ac8-11eb-92a6-2b3fd86d976e.png)
![Screenshot 2020-12-10 at 09 06 26](https://user-images.githubusercontent.com/22518997/101740018-9dc90800-3ac8-11eb-9b31-5b84572e2397.png)

Loading kubernetes nodes
![Uploading Screenshot 2020-12-10 at 09.06.05.png…]()

Loading services
![Screenshot 2020-12-10 at 09 09 58](https://user-images.githubusercontent.com/22518997/101740097-b6392280-3ac8-11eb-91d8-212659f5b76d.png)

Loading pods logs
![Screenshot 2020-12-10 at 09 06 42](https://user-images.githubusercontent.com/22518997/101740047-a7eb0680-3ac8-11eb-829c-2ebca1891b4f.png)

Weird CRDs menu behaviour
![Screenshot 2020-12-10 at 09 07 15](https://user-images.githubusercontent.com/22518997/101740129-bf29f400-3ac8-11eb-968c-e22a783dccef.png)
![Screenshot 2020-12-10 at 09 07 21](https://user-images.githubusercontent.com/22518997/101740131-bfc28a80-3ac8-11eb-8a2d-0c24d34181e0.png)
![Screenshot 2020-12-10 at 09 07 25](https://user-images.githubusercontent.com/22518997/101740133-bfc28a80-3ac8-11eb-9159-74a8be188ca1.png)

**Environment (please complete the following information):**
- Lens Version: since 4.x
- OS: MacOS Catalina 10.15.7
- Installation method: dmg for old versions, auto-upgrade to 4.x
- Kubernetes versions: 1.17.4 (prod) 1.14 (others)
- Cloud provider: GCP/GKE private cluster
- Permissions: k8s cluster admin

**Logs:**
When you run the application executable from command line you will see some logging output. Please paste them here:
```
info: 🚀 Starting Lens from "/Users/mbaiguini/Library/Application Support/Lens"
info: Checking for update
info: [STORE]: LOADED from /Users/mbaiguini/Library/Application Support/Lens/config.json
info: [STORE]: LOADED from /Users/mbaiguini/Library/Application Support/Lens/lens-cluster-store.json
info: [STORE]: LOADED from /Users/mbaiguini/Library/Application Support/Lens/lens-workspace-store.json
info: [STORE]: LOADED from /Users/mbaiguini/Library/Application Support/Lens/lens-extensions.json
info: [STORE]: LOADED from /Users/mbaiguini/Library/Application Support/Lens/lens-filesystem-provisioner-store.json
info: [CLUSTER-MANAGER]: init cluster {"id":"fa9c4407-aa2e-4c31-a2ed-3af63ef99571","name":"dev","initialized":false,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: [CLUSTER-MANAGER]: init cluster {"id":"82a7af51-7773-4ef8-9877-20fb1eb2f6bb","name":"test","initialized":false,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: [CLUSTER-MANAGER]: init cluster {"id":"efce6e5a-b208-459c-b6c0-c1c9d70fdfeb","name":"security-tests","initialized":false,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: [CLUSTER-MANAGER]: init cluster {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":false,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: LensProxy server has started at http://localhost:55463
info: [EXTENSION-DISCOVERY] watching extension add/remove in /Users/mbaiguini/.k8slens/extensions
info: [EXTENSION-DISCOVERY] loading extensions from /Users/mbaiguini/Library/Application Support/Lens
info: [CLUSTER-MANAGER]: init cluster {"id":"82a7af51-7773-4ef8-9877-20fb1eb2f6bb","name":"test","initialized":false,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: [CLUSTER-MANAGER]: init cluster {"id":"efce6e5a-b208-459c-b6c0-c1c9d70fdfeb","name":"security-tests","initialized":false,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: [CLUSTER-MANAGER]: init cluster {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":false,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: [CLUSTER]: "dev" init success {"id":"fa9c4407-aa2e-4c31-a2ed-3af63ef99571","context":"dev","apiUrl":"https://35.233.42.230"}
info: [CLUSTER-MANAGER]: init cluster {"id":"efce6e5a-b208-459c-b6c0-c1c9d70fdfeb","name":"security-tests","initialized":false,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: [CLUSTER-MANAGER]: init cluster {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":false,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: [CLUSTER]: "test" init success {"id":"82a7af51-7773-4ef8-9877-20fb1eb2f6bb","context":"test","apiUrl":"https://35.246.7.249"}
info: [CLUSTER-MANAGER]: init cluster {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":false,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: [CLUSTER]: "security-tests" init success {"id":"efce6e5a-b208-459c-b6c0-c1c9d70fdfeb","context":"security-tests","apiUrl":"https://35.246.100.198"}
info: [CLUSTER]: "PROD" init success {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","context":"PROD","apiUrl":"https://34.65.163.184"}
info: [CLUSTER]: "test" init success {"id":"82a7af51-7773-4ef8-9877-20fb1eb2f6bb","context":"test","apiUrl":"https://35.246.7.249"}
info: [CLUSTER]: "security-tests" init success {"id":"efce6e5a-b208-459c-b6c0-c1c9d70fdfeb","context":"security-tests","apiUrl":"https://35.246.100.198"}
info: [CLUSTER]: "PROD" init success {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","context":"PROD","apiUrl":"https://34.65.163.184"}
info: [CLUSTER]: "security-tests" init success {"id":"efce6e5a-b208-459c-b6c0-c1c9d70fdfeb","context":"security-tests","apiUrl":"https://35.246.100.198"}
info: [CLUSTER]: "PROD" init success {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","context":"PROD","apiUrl":"https://34.65.163.184"}
info: [CLUSTER]: "PROD" init success {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","context":"PROD","apiUrl":"https://34.65.163.184"}
info: [EXTENSION-INSTALLER] installing dependencies at /Users/mbaiguini/Library/Application Support/Lens
info: Update for version 4.0.1 is not available (latest version: 4.0.1, downgrade is disallowed).
info: [EXTENSION]: enabled lens-license@0.1.0
telemetry main extension activated
info: [STORE]: LOADED from /Users/mbaiguini/Library/Application Support/Lens/extension-store/lens-telemetry/preferences-store.json
info: [EXTENSION]: enabled lens-telemetry@0.1.0
info: [CLUSTER]: activate {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: [CLUSTER]: bind events {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: [CLUSTER]: reconnect {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":false,"online":false,"accessible":false,"disconnected":true}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"Authentication proxy started\n","meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":false,"online":false,"accessible":false,"disconnected":true}}
warn: PrometheusLens: failed to list services: services "prometheus" not found
warn: PrometheusLens: failed to list services: services "prometheus" not found
warn: PrometheusLens: failed to list services: services "prometheus" not found
warn: PrometheusLens: failed to list services: services "prometheus-server" not found
warn: PrometheusLens: failed to list services: services "prometheus-server" not found
warn: PrometheusLens: failed to list services: services "prometheus-server" not found
warn: PrometheusLens: failed to list services: services "prometheus" not found
info: using operator as prometheus provider
warn: PrometheusLens: failed to list services: services "prometheus-server" not found
info: using operator as prometheus provider
(node:45418) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
info: [CLUSTER]: refresh {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:07 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n2020/12/10 09:30:07 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:19 http: proxy error: context canceled\n2020/12/10 09:30:19 http: proxy error: context canceled\n2020/12/10 09:30:19 http: proxy error: context canceled\n2020/12/10 09:30:19 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:19 http: proxy error: context canceled\n2020/12/10 09:30:19 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [CLUSTER]: refresh {"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:38 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:38 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:38 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n2020/12/10 09:30:38 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:49 http: proxy error: context canceled\n2020/12/10 09:30:49 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:49 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:49 http: proxy error: context canceled\n2020/12/10 09:30:49 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:49 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:50 http: proxy error: context canceled\n2020/12/10 09:30:50 http: proxy error: context canceled\n2020/12/10 09:30:50 http: proxy error: context canceled\n2020/12/10 09:30:50 http: proxy error: context canceled\n2020/12/10 09:30:50 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
info: [KUBE-AUTH]: out-channel "kube-auth:ff465b34-513b-4031-bb4e-af669ee55f5f" {"data":"2020/12/10 09:30:50 http: proxy error: context canceled\n","error":true,"meta":{"id":"ff465b34-513b-4031-bb4e-af669ee55f5f","name":"PROD","initialized":true,"ready":true,"online":true,"accessible":true,"disconnected":false}}
```

**Kubeconfig:**
Quite often the problems are caused by malformed kubeconfig which the application tries to load. Please share your kubeconfig, remember to remove any secret and sensitive information.
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: ...
    server: ...
  name: gke_prod-cluster
- cluster:
    certificate-authority-data: ...
    server: ...
  name: gke_dev-cluster
- cluster:
    certificate-authority-data: ...
    server: ...
  name: gke_security-cluster
- cluster:
    certificate-authority-data: ...
    server: ...
  name: gke_test-cluster
contexts:
- context:
    cluster: gke_prod-cluster
    namespace: tee
    user: gke_prod-cluster
  name: PROD
- context:
    cluster: gke_dev-cluster
    namespace: monitoring
    user: gke_dev-cluster
  name: dev
- context:
    cluster: gke_security-cluster
    namespace: monitoring
    user: gke_security-cluster
  name: security
- context:
    cluster: gke_test-cluster
    namespace: monitoring
    user: gke_test-cluster
  name: test
current-context: PROD
kind: Config
preferences: {}
users:
- name: gke_prod-cluster
  user:
    auth-provider:
      config:
        ...
      name: gcp
- name: gke_dev-cluster
  user:
    auth-provider:
      config:
        ...
      name: gcp
- name: gke_security-cluster
  user:
    auth-provider:
      config:
        ...
      name: gcp
- name: gke_test-cluster
  user:
    auth-provider:
      config:
        ...
      name: gcp
```

Please help!! Lens is really important for our everyday job!

**Comment by myspotontheweb** [comment]
Duplicate?

https://github.com/lensapp/lens/issues/1700

Perhaps this [work-around](https://github.com/lensapp/lens/issues/1700#issuecomment-741856445) might help

**Comment by bygui86** [comment]
@myspotontheweb let me check if I have same errors in the console...
anyway before closing this, please merge the two issues as I put some useful info and the other guy as well

**Comment by bygui86** [comment]
@myspotontheweb 

> Perhaps this work-around might help

this workaround does not help, I have cluster-admin role and full access to whole gcp/gke/k8s

**Comment by bygui86** [comment]
@myspotontheweb I think I have a different issue...

here the console right after opening a cluster
![Screenshot 2020-12-10 at 11 43 06](https://user-images.githubusercontent.com/22518997/101763602-4b471600-3adf-11eb-99a8-925fdae21fbe.png)

here network right after opening a cluster
![Screenshot 2020-12-10 at 11 43 48](https://user-images.githubusercontent.com/22518997/101763634-569a4180-3adf-11eb-9c66-478c504f58e9.png)

and here network after a while the cluster was opened
![Screenshot 2020-12-10 at 11 45 39](https://user-images.githubusercontent.com/22518997/101763652-5ef27c80-3adf-11eb-9e73-20a334f71ed2.png)

let me know if you need some more info

**Comment by bygui86** [comment]
@myspotontheweb for sure this bug is related to #1700, but not a duplicate!

We have currently 3 clusters on GCP: `dev`, `test` and `prod`.

`dev` and `test` are running k8s 1.15.x and are normal clusters
`prod` is running k8s 1.17.14 and is a **GKE private cluster**

My issue is related only to the `prod` private cluster!!

GKE private cluster means that k8s worker nodes are not exposed to internet, but are available only internally. K8s master node is normally exposed like all others. This means that tools like `kubectl` and `k9s` are able to reach that cluster normally without issues...

Do you know why Lens has issues with GKE private clusters?

**Comment by matteovivona** [comment]
@bygui86 same issue here. I have thought that it was some particular port blocked by GCP firewall, but I haven't found any evidence

**Comment by zifeo** [comment]
@bygui86 Not sure how the private GKE issue relates with Lens version 4.0 (might the introduction of the `kubectl proxy` usage, but unfamiliar with the project and its architecture), but this seems to be a limitation of vanilla GKE private clusters.

```
DEBUG_PROXY=true /Applications/Lens.app/Contents/MacOS/Lens
# shows timeout while querying http://K8SMASTERIP/api/v1/namespaces/NAMESPACE/services/PROMETHEUSSERVICE:9090/proxy/api/v1/query_range?query=PROMETHEUSQUERY
```

This can be reproduced without Lens (ensure you are using the correct cluster context):
```
kubectl proxy
# Starting to serve on 127.0.0.1:8001
curl http://localhost:8001/api/v1/namespaces/NAMESPACE/services/PROMETHEUSSERVICE:9090/proxy/api/v1/query_range?query=PROMETHEUSQUERY
# Error trying to reach service: 'dial tcp LOCALIP:9090: i/o timeout'
```

This timeout is caused by the VPC firewall rule between the masters and the other nodes (rule name ending with XXXX-master and ports 10250 & 443 by default). As often required by [admission controllers](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/), some ports need to be opened there. In our case, adding port 9090 solves the issue, beware this might affect cluster isolation as well.

**Comment by bygui86** [comment]
@zifeo Thanks for the explanation!

Apart from the Prometheus metrics missing issue, the rest of issues were solved with latest upgrades (> 4.0.2, currently I have 4.0.8).

I checked and I found the same error message as you wrote above: it seems that Prometheus is not reachable.
The missing of firewall rule makes sense to me and due to security reasons I cannot open port 9090, so fair enough :) I will use Lens without Prometheus metrics from GKE private clusters for now.

I know `kubectl proxy` is pretty common and powerful, but may I suggest to add the possibility to use `kubectl port-forward` instead? Even only for Prometheus metrics would be great!
This way opening a port on firewall can be avoided and security posture would stay on the safe side.

**Comment by pslobo** [comment]
Just want to chime in here as the issue #2009 seems to be related. While I do have permissions to open port 9090 for testing to confirm this "solves" the issue, we can't leave it open. So a port-forward would be better for this particular use case.

**Comment by pslobo** [comment]
> I will use Lens without Prometheus metrics from GKE private clusters for now.

@bygui86 downside of this is that Lens becomes slow, almost unusable while it tries to load metrics it can't. Can you confirm this happens to you too?

**Comment by zifeo** [comment]
@pslobo @bygui86 Could you explain why opening this port is a no-go for your org? Port-forward would definitely be a better way to handle that, however curious to see your risk model behind that.

**Comment by pslobo** [comment]
@zifeo not the one that defines security policies, but from where I stand, most security concerns around this can be mitigated with just allowing master authorized networks access to port 9090 for non-prod clusters. For prod clusters, I can straight up say the security team will most likely not allow this (then again, we really shouldn't be mucking around too much in prod 😁)

I'd have to say (guessing here), that the security team wants to reduce the attack vector as much as possible and therefore, having an exposed port that could possibly be exploited in some way that we're not yet aware of, is a no-go for them.

**Comment by bygui86** [comment]
@pslobo same here. Lens is very slow and sometime unresponsive for some seconds trying to load metrics.

@zifeo in our company we follow least privileges principle and smallest attack surface possible. That's why Prometheus runs in an isolated namespace with the minimum amount of privileges inside the Kubernetes cluster, but I still agree with @pslobo saying that it's not acceptable to be forced to open a well-known port on a production cluster without a business/functional reason to justify it. Considering that during debugging normally Grafana is used and not directly Prometheus.
Moreover Prometheus has no login mechanism (unless for example behind a reverse proxy or GCP Identity-Access Proxy), so it would be completely exposed, offering an attack vector.

Imo it's the same as having a GKE public cluster instead of a private one: public cluster is already secure because Google grants it, but why not a better security level offered by private cluster?

**Comment by zifeo** [comment]
@bygui86 @pslobo Wanted to see if you see non-obvious treat. I agree that is not best practice and should be avoided 👍

**Comment by Nokel81** [comment]
So if I am understanding correctly, this issue has to do with not being able to proxy to your cluster or is it that Lens can connect but doesn't display any info.

**Comment by bygui86** [comment]
@Nokel81 actually Lens is not able to communicate properly with GKE private clusters. Lens is slower and does not display any additional info (cpu and memory usage for example) as it's not able to find Prometheus deoyed on the cluster.

**Comment by Nokel81** [comment]
But you are able to connect to that GKE cluster using `kubectl`?

**Comment by bygui86** [comment]
Yeah no limitations on that.
Admin/owner permissions.
No restrictions on namespaces.
Prometheus reachable using kubectl port forwarding.

**Comment by terrpan** [comment]
I have the same issue as @bygui86 but on Linode. Everything works except metrics. And after a while the whole thing gets unusable and I have to reload or sometimes even restart Lens to get any updates.



In the logs I get a lot of these messages:
```
info: [KUBE-AUTH]: out-channel "kube-auth:6333a98ad8dccb1561f324508064a8c1" {"data":"2021/07/08 09:49:44 http: proxy error: context canceled\n","error":true,"meta":{"id":"6333a98ad8dccb1561f324508064a8c1","name":"REDACTED","ready":false,"online":false,"accessible":false,"disconnected":true}}
```

and after some time the log gets spammed with these:
```
error: [Metrics]: metrics not available {"error":{"name":"RequestError","message":"Error: socket hang up","cause":{"code":"ECONNRESET"},"error":{"code":"ECONNRESET"},"options":{"timeout":0,"resolveWithFullResponse":false,"json":true,"method":"POST","form":{"query":"sum(kube_pod_container_resource_limits{pod=~\"whoami-796959c99b-fqp4g\",resource=\"cpu\",namespace=\"default\"}) by (container, namespace)","start":"1625726880","end":"1625730480","step":"60","kubernetes_namespace":"default"},"headers":{"Host":"6333a98ad8dccb1561f324508064a8c1.localhost:62816"},"uri":"http://localhost:62816/api-kube/api/v1/namespaces/monitoring/services/prometheus-operated:9090/proxy/api/v1/query_range","simple":true,"transform2xxOnly":false}}}
```

Lens: Lens: 5.0.2-latest.20210705.2

**Comment by cedricve** [comment]
@terrpan tried my linode clusters, and I do not get any results at all. I just see the available nodes, but that's it. Did you find any solution for this? I believe it's some kind of cluster role missing.

**Comment by lucaslorentz** [comment]
For those who don't know how how to do what @zifeo suggested.

Find gke firewall id (replace CLUSTER_NAME by your cluster name):
```
gcloud compute firewall-rules list --filter="name~gke-${CLUSTER_NAME}-[0-9a-z]*-master"
```

Add more ports to the found firewall rule (port 9090 to get Lens working properly):
```
gcloud compute firewall-rules update <firewall-rule-name> --allow tcp:10250,tcp:443,tcp:9090
```

Instructions were adapted from https://istio.io/latest/docs/setup/platform-setup/gke/ , in my case I added ports 15017 for Istio and 9090 for Lens Prometheus.

**Comment by Dhanabal2710** [comment]
How did you people use lens when connecting private GKE clusters via IAP identity access proxy?

**Comment by DanTulovsky** [comment]
I am also unable to connect to a GKE private cluster using OpenLens.  That is, it connects, but I cannot see any workloads.  It does see namespaces and CRDs, as well as helm releases.  Everything else is blank.

I tried the latest alpha version as well as the stable one, and even 6.3.0.

The caveat is the Lens IDE (the paid version) does work just fine against the same cluster.

Happy to provide any more info to help fix this.

Lens version (working): 2023.1.110749-latest
OpenLens version (not working): > 6.3.0. (didn't try anything before that)

**Comment by Nokel81** [comment]
@DanTulovsky Your issue should be fixed by #6995
